File: ccms/src/svc/goose/cms_set_go_cb_result.c
#include "svc/goose/cms_set_go_cb_result.h"
#include "data/scalar/cms_boolean.h"
#include "data/common/cms_service_error.h"
#include "per/cms_sequence.h"

int cms_set_go_cb_result_encode_stream(per_stream_t *s, const cms_set_go_cb_result_t *v) {
    if (!v)
        return CMS_ERR;
    int err;

    /* 0. OPTIONAL bitmap (4 fields: error, goEnaErr, goIdErr, datSetErr) */
    bool opt_present[4] = {(v->error_present && v->error_present->value) && v->error,
                           (v->go_ena_err_present && v->go_ena_err_present->value) && v->go_ena_err,
                           (v->go_id_err_present && v->go_id_err_present->value) && v->go_id_err,
                           (v->dat_set_err_present && v->dat_set_err_present->value) && v->dat_set_err};
    err = (int) per_encode_optional_bitmap(s, opt_present, 4);
    if (err)
        return err;

    /* 1. error — ServiceError OPTIONAL (bitmap[0]) */
    if (opt_present[0]) {
        err = cms_service_error_encode_stream(s, v->error);
        if (err)
            return err;
    }

    /* 2. goEnaErr — ServiceError OPTIONAL (bitmap[1]) */
    if (opt_present[1]) {
        err = cms_service_error_encode_stream(s, v->go_ena_err);
        if (err)
            return err;
    }

    /* 3. goIdErr — ServiceError OPTIONAL (bitmap[2]) */
    if (opt_present[2]) {
        err = cms_service_error_encode_stream(s, v->go_id_err);
        if (err)
            return err;
    }

    /* 4. datSetErr — ServiceError OPTIONAL (bitmap[3]) */
    if (opt_present[3]) {
        err = cms_service_error_encode_stream(s, v->dat_set_err);
        if (err)
            return err;
    }

    return CMS_OK;
}

int cms_set_go_cb_result_decode_stream(per_stream_t *s, void *ptr) {
    cms_set_go_cb_result_t *v = (cms_set_go_cb_result_t *) ptr;
    int err;

    /* 0. OPTIONAL bitmap (4 fields) */
    bool opt_present[4];
    err = (int) per_decode_optional_bitmap(s, opt_present, 4);
    if (err)
        return err;
    if (v) {
        if (v->error_present)
            v->error_present->value = opt_present[0];
        if (v->go_ena_err_present)
            v->go_ena_err_present->value = opt_present[1];
        if (v->go_id_err_present)
            v->go_id_err_present->value = opt_present[2];
        if (v->dat_set_err_present)
            v->dat_set_err_present->value = opt_present[3];
    }

    /* 1. error OPTIONAL */ if (opt_present[0]) {
        if (v && !v->error)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->error : NULL);
        if (err)
            return err;
    }
    /* 2. goEnaErr OPTIONAL */ if (opt_present[1]) {
        if (v && !v->go_ena_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->go_ena_err : NULL);
        if (err)
            return err;
    }
    /* 3. goIdErr OPTIONAL */ if (opt_present[2]) {
        if (v && !v->go_id_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->go_id_err : NULL);
        if (err)
            return err;
    }
    /* 4. datSetErr OPTIONAL */ if (opt_present[3]) {
        if (v && !v->dat_set_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->dat_set_err : NULL);
        if (err)
            return err;
    }

    return CMS_OK;
}

File: jcms/jcms-core/src/main/java/com/ysh/jcms/svc/goose/CmsSetGoCbResult.java
package com.ysh.jcms.svc.goose;

import com.ysh.jcms.core.CmsType;
import com.ysh.jcms.data.common.CmsServiceError;
import com.ysh.jcms.data.scalar.CmsBoolean;
import java.util.Arrays;
import java.util.List;

/**
 * SetGoCBResult ::= SEQUENCE { error [0] IMPLICIT ServiceError OPTIONAL, goEna
 * [1] IMPLICIT ServiceError OPTIONAL, goID [2] IMPLICIT ServiceError OPTIONAL,
 * datSet [3] IMPLICIT ServiceError OPTIONAL } — 8.9.5
 *
 * Used by SetGoCBValues error.
 */
public class CmsSetGoCbResult extends CmsType {

    public CmsBoolean errorPresent;
    public CmsServiceError error;
    public CmsBoolean goEnaErrPresent;
    public CmsServiceError goEnaErr;
    public CmsBoolean goIdErrPresent;
    public CmsServiceError goIdErr;
    public CmsBoolean datSetErrPresent;
    public CmsServiceError datSetErr;

    public CmsSetGoCbResult() {
        this.errorPresent = new CmsBoolean();
        this.error = new CmsServiceError();
        this.goEnaErrPresent = new CmsBoolean();
        this.goEnaErr = new CmsServiceError();
        this.goIdErrPresent = new CmsBoolean();
        this.goIdErr = new CmsServiceError();
        this.datSetErrPresent = new CmsBoolean();
        this.datSetErr = new CmsServiceError();
    }

    public CmsSetGoCbResult errorPresent(boolean v) {
        this.errorPresent.value(v);
        return this;
    }
    public CmsSetGoCbResult error(int v) {
        this.error.value(v);
        return this;
    }
    public CmsSetGoCbResult goEnaErrPresent(boolean v) {
        this.goEnaErrPresent.value(v);
        return this;
    }
    public CmsSetGoCbResult goEnaErr(int v) {
        this.goEnaErr.value(v);
        return this;
    }
    public CmsSetGoCbResult goIdErrPresent(boolean v) {
        this.goIdErrPresent.value(v);
        return this;
    }
    public CmsSetGoCbResult goIdErr(int v) {
        this.goIdErr.value(v);
        return this;
    }
    public CmsSetGoCbResult datSetErrPresent(boolean v) {
        this.datSetErrPresent.value(v);
        return this;
    }
    public CmsSetGoCbResult datSetErr(int v) {
        this.datSetErr.value(v);
        return this;
    }

    @Override
    public List<? extends CmsType> children() {
        return Arrays.asList(errorPresent, error, goEnaErrPresent, goEnaErr, goIdErrPresent, goIdErr, datSetErrPresent, datSetErr);
    }
}
